atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/compact/BackgroundCompactor.java

```diff
@@ -0,0 +1,184 @@
+
+                    registerCompactedTable(tableToCompact);
+                }
+            }
+        } catch (InterruptedException e) {
+            log.warn("Shutting down background compactor due to InterruptedException", e);
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    private SingleLockService createSimpleLocks() {
+        return new SingleLockService(lockService, "atlas compact");
+    }
+
+    private void registerCompactedTable(String tableToCompact) {
+        transactionManager.runTaskWithRetry(tx -> {
+            CompactMetadataTable compactMetadataTable = CompactTableFactory.of().getCompactMetadataTable(tx);
+            compactMetadataTable.putLastCompactTime(
+                    CompactMetadataTable.CompactMetadataRow.of(tableToCompact),
+                    System.currentTimeMillis());
+
+            return null;
+        });
+    }
+
+    private void compactTable(String tableToCompact) {
+        keyValueService.compactInternally(TableReference.createFromFullyQualifiedName(tableToCompact));
+    }
+
+    private Optional<String> selectTableToCompact(Transaction tx) {
+        Map<String, Long> tableToLastTimeSwept = new HashMap<>();
+        SweepPriorityTable sweepPriorityTable = SweepTableFactory.of().getSweepPriorityTable(tx);
+        sweepPriorityTable.getAllRowsUnordered(SweepPriorityTable.getColumnSelection(
+                SweepPriorityTable.SweepPriorityNamedColumn.LAST_SWEEP_TIME))
+                .forEach(row -> {
+                    Long lastSweepTime = row.getLastSweepTime();
+                    String tableName = row.getRowName().getFullTableName();
+                    tableToLastTimeSwept.put(tableName, lastSweepTime);
+                });
+
+        Map<String, Long> tableToLastTimeCompacted = new HashMap<>();
+        CompactMetadataTable compactMetadataTable = CompactTableFactory.of().getCompactMetadataTable(tx);
+        compactMetadataTable.getAllRowsUnordered(SweepPriorityTable.getColumnSelection(
+                SweepPriorityTable.SweepPriorityNamedColumn.LAST_SWEEP_TIME))
+                .forEach(row -> {
+                    Long lastCompactTime = row.getLastCompactTime();
+                    String tableName = row.getRowName().getFullTableName();
+                    tableToLastTimeCompacted.put(tableName, lastCompactTime);
+                });
+
+        List<String> uncompactedTables = tableToLastTimeSwept.keySet().stream()
+                .filter(table -> !tableToLastTimeCompacted.keySet().contains(table))
+                .collect(Collectors.toList());
+
+        if (uncompactedTables.size() > 0) {
+            int randomTableIndex = ThreadLocalRandom.current().nextInt(uncompactedTables.size());
+            return Optional.of(uncompactedTables.get(randomTableIndex));
+        }
+
+        String tableToCompact = null;
+        long maxSweptAfterCompact = Long.MIN_VALUE;
```

Comment:
This should be set to 0, since we don't want to compact anything if all tables have been compacted since they were last swept.